src/status_im/contexts/wallet/events.cljs

```diff
@@ -704,14 +704,19 @@
 (rf/reg-event-fx
  :wallet/sign-transactions-signal-received
  (fn [{:keys [db]} [{send-details :sendDetails :as data}]]
-   (let [type (if (or (= (:fromToken send-details)
-                         (:toToken send-details))
-                      (string/blank? (:toToken send-details)))
-                :send
-                :swap)]
-     (if-let [error-response (:errorResponse send-details)]
-       {:db (assoc-in db [:wallet :ui type :error-response] error-response)}
-       {:db (assoc-in db [:wallet :ui type :transaction-for-signing] data)}))))
+   (let [type           (if (or (= (:fromToken send-details)
+                                   (:toToken send-details))
+                                (string/blank? (:toToken send-details)))
+                          :send
+                          :swap)
+         callback-fx    (get-in db [:wallet :ui type :sign-transactions-callback-fx])
+         error-response (:errorResponse send-details)]
+     {:fx [(when (and callback-fx (not error-response))
+             callback-fx)]
```

Comment:
this is a weird pattern. can't we just delay the showing of the modal instead?

Reply:
Why? I'd say delaying showing of the modal is a weird pattern because it means delaying it to certain amount of time instead of doing it immediately after you've got a response.

Reply:
I mean the pattern of storing a callback in the db specifically. Does it solve any particular issue/bug? Asking cause it seems odd and I wonder if it can be done otherwise?

cc: @ilmotta @vkjr wdyt?

Reply:
As an alternative to passing the callback, why don't we just dispatch the show-modal from here directly? When we dispatch the `:wallet/build-transactions-from-route` event, we can also specify the kind of transaction it is e.g. `transfer`/`swap`/`approval`/`bridge`/etc. and store it in the db. Then when we get the signal, we can react to it conditionally i.e. by showing the modal in case of `approval`. IMO this makes the steps of the flow clearer to read/understand, but maybe it's just me.

P.S. In general I don't get why this even has to be a signal, as it completely detaches the build transaction "request" from the "response", so we end up populating the db with intermediary state just to understand in which context the request was called in. Just getting the response from the rpc directly would've been so much easier IMO. @ilmotta from last week's discussion, I think this is an example where signals get in the way a bit.

Reply:
The naming is a bit confusing to me because it's not really a callback, but an effect passed to `:fx`. Callbacks in my head imply being callable as functions. I'd use the name `:sign-transactions-fx` without the `callback` word.

----

> P.S. In general I don't get why this even has to be a signal, as it completely detaches the build transaction "request" from the "response", so we end up populating the db with intermediary state just to understand in which context the request was called in. Just getting the response from the rpc directly would've been so much easier IMO. @ilmotta from last week's discussion, I think this is an example where signals get in the way a bit.

But as @clauxx said, I agree, we should ask ourselves why go the signal approach instead of the simpler req/res. I'm not saying generally, specifically about `BuildTransactionsFromRoute` endpoint.

Signals are great to keep state in sync between client and status-go. As a comment in status-go explains, perhaps the reason for using signals is because the endpoint has to notify the client of two steps `RouterSendingTransactionsStarted` and `SignRouterTransactions`. Is this really necessary to build the feature in the app?

Go is quite efficient in handling concurrent IO, so I'd imagine building the transactions, even if it takes dozens of seconds in worst case scenarios wouldn't be a problem at all for other goroutines in status-go to keep doing their job.

I lack context, but it would be great to bring this discussion @clauxx to a status-go channel or maybe a call with @saledjenic, @igor-sirotin, @osmaczko, the whole gang because this "pub/sub vs req/res" discussion is almost as old as software development. Not to answer what's best generally, which would be shortsighted, but to evaluate some cases, such as this `BuildTransactionsFromRoute`.